### DISCUSSION TITLE
Want to go ahead with Airflow Production, need guidance on parameters. Completed POC

### DISCUSSION THREAD

**Discussion by AustinJackson-GitHub** [apache/airflow#30574] [Q&A]
We needed input for the Airflow production setup, as we were able to accomplish a small POC on SQLite and a sequential executor as well as local executor in Airflow 2.5.0.

    Well, we have almost 10 concurrent jobs, which we want to schedule in airflow.
    We have almost 70 jobs in total, that will be totally pushed onto the system one after another.
    Previously, our jobs were on Jenkins, which is used to schedule + execute these jobs. But, now we will go ahead with Airflow to schedule
    Over and above, these all need to be automated for scheduling, whereas these jobs would be using the Docker instance for different projects with individual DAG's as their jobs for daily purposes.
    We want to go-ahead with the AWS Ec2 instance as an airflow server with MySQL as our metadata back-end.
    Well, we need input on the parameters: 6.a Parameters of airflow.cfg 6.b On infrastructure 6.c Any good practices 6.d Thumb rules 6.e Good experience 6.f Industry standard

Please guide us on front of both s/w and h/w including the parameters required keeping above things into consideration

NOTE: We don't want to go onto Kubernetes currently

**Comment by potiuk** [comment]
It's impossible to come up with such guidance. 

Airflow is a complex and distributed system and moreover, it runs user-provided code of arbitrary complexity. whatever you do in your DAGs might impact performance in multiple ways and hardware configuration so there are no "industry" standards or configuration and it is impossible to even guess what will work in your case.

However you did well - you run a POC and this is the right approach that you should continue - build your system with some spare capacity starting from minimum size that your testing will see can run without any of your code (depending on the choice of executors/distribution mechanism for DAGs, Cloud, VM/containers etc. those might be different "minima" but generally our simple "development instances" start from 2 CPU and 4 GB RAM in general (this is know to work for "empty airflow" with local executor.

Then you can add your DAGs and assess and monitor the usage of resources and see if you need to grow them as your DAG base grows. 

Generally scaling and foreseeing scale of systems in cloud/distributed environment is very different than it used to be 20 or even 10 years ago - the number of layers and complexity of depenencies makes it almost impossieble to foresse and accurately estimate the right size of hardware you need. This is extremely complex to foresee even for simple system and requires extensive benchmarking (and airflow is run by volunteers and has many knobs to turn to tune, and is highly distributed system in general, so it's impossible to have such benchmarks even if we would limit it to just a few "basic" configurations of all the "moving parts".

Instead the whole industry moved already from "plan and foresee" to "observe and adjust" model  - especially in the cloud which is prepared for any scaling events you might imagine, and the best practice is to deploy your system in the way that you can easily failover to any other hardware (which happens all the time btw so you have to do it).

In "process control" theory this is really distinguishes highly-predictable systems where you can foresee everything and calculate upfront (happens in civil engineering for obvious reasons and comes usually with 60%-80% margins of calculations added for safety) with systems that has so many moving parts that it makes it costly and nearly impossible to calculate everything and it relies more on a feedback loop where you observe system and adjust (which is much better for utilization of resources over time as you can adjust as you go).

With Airflow, I am afraid you need to go with the "feedback" loop approach. Start smalll and add resources as you grow.

**Reply by AustinJackson-GitHub** [reply]
Thank you for the update Potiuk, needed guidance for the HA configurations and its implementation method. Would be very appreciable if any document for rerference.

**Reply by potiuk** [reply]
Just the standard installation docs https://airflow.apache.org/docs/apache-airflow/stable/installation/index.html - you can see all the installation options you have with pros and cons so that you can analyse them. 

You can look at those (and our chart https://airflow.apache.org/docs/helm-chart/stable/index.html ) to see soem of the ways how you can do deployment. We recommned Helm Chart and Kubernetes for Production use (this is the only deployment where you will get more details and Production Guide, and even if (as you wrote) you do not wish to use it, you can learn a lot from it. When you run `helm install --dry-run` with various options (including scalability, monitoring etc.) on our chart you will be able to see what community approach for production-ready Helm chart is.

This is as good of a (self) guidance as you can get here. But there are various community resources you can brows (for example Ecosystem https://airflow.apache.org/ecosystem/ page of ours has a number of links, talks, blogs linked that might better suite your case.

**Reply by AustinJackson-GitHub** [reply]
Thanks a lot and sincerely appreciate the help.
